### Assignment 11.2

#### Build the linear regression model using scikit learn in boston data to predict 'Price' based on other dependent variable.

In [65]:
import sqlite3
import numpy as np
import pylab as pl
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt

from math import sqrt
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, cross_val_score

%matplotlib inline

<b>Load the Boston Data set. Adding the Column names and Target values into the DataFrame.</b>

In [66]:
from sklearn.datasets import load_boston
boston = load_boston()
bos = pd.DataFrame(boston.data)

In [67]:
bos.shape

(506, 13)

In [15]:
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR'])

In [70]:
bos.columns = boston.feature_names

In [72]:
bos['TARGET'] = boston.target

In [73]:
bos.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  TARGET  
0     15.3  396.90   4.98    24.0  
1     17.8  396.90   9.14    21.6  
2     17.8  392.83   4.03    34.7  
3     18.7  394.63   2.94    33.4  
4     18.7  396.90   5.33    36.2

<b>Data Cleaning and EDA</b>

In [74]:
bos.shape

(506, 14)

In [75]:
bos.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.593761   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.596783   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.647423   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT      TARGET  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

In [76]:
print(boston.DESCR)

Boston House Prices dataset

Notes
------
Data Set Characteristics:  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive
    
    :Median Value (attribute 14) is usually the target

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pupil-teacher ratio by town
      

<b>Train and Fit the Linear Regression Model</b>

In [85]:
X = bos.drop('TARGET', axis = 1)
y = bos['TARGET']

In [89]:
X.shape

(506, 13)

In [90]:
y.shape

(506,)

In [91]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.3, random_state = 22)

In [92]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(354, 13)
(152, 13)
(354,)
(152,)


In [93]:
lm = LinearRegression()

In [94]:
lm.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [95]:
y_pred = lm.predict(X_test)

In [96]:
np.sqrt(mean_squared_error(y_test,y_pred))

4.806965665002193

In [132]:
Result = pd.DataFrame(X.columns, columns = ['Coefficients'])
Result['Value'] = lm.coef_
df = pd.DataFrame([['INTERCEPT', lm.intercept_]], columns = ['Coefficients', 'Value'])
Result.append(df, ignore_index = True) 

Coefficients      Value
0          CRIM  -0.100698
1            ZN   0.041850
2         INDUS  -0.011446
3          CHAS   3.503657
4           NOX -14.800825
5            RM   3.527116
6           AGE  -0.000471
7           DIS  -1.423916
8           RAD   0.269763
9           TAX  -0.009742
10      PTRATIO  -0.807562
11            B   0.011347
12        LSTAT  -0.566073
13    INTERCEPT  33.299967

<b>K-Fold cross validation</b>

In [97]:
scores = cross_val_score(LinearRegression(), X, y, scoring='neg_mean_squared_error', cv=10)
scores, scores.mean()

(array([  -9.30130569,  -14.17208374,  -14.10631149,  -35.19479431,
         -31.92474931,  -19.85853694,   -9.95792   , -168.53075207,
         -33.61791833,  -10.96654317]), -34.76309150542213)

In [63]:
np.sqrt(scores.mean() * -1)

5.8960233637106745

<b>Interpreting the Coefficients and its Values:</b>


e.g. NOX and Target:
<br>Assuming all the values being constant, for a unit increase in NOX, the target value decreases by 14.800825 units.